semantic-analysis/src/Analysis/Name.hs

```diff
@@ -1,40 +1,72 @@
-{-# LANGUAGE DeriveTraversable, GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
 module Analysis.Name
-( Name(..)
-, Named(..)
-, named
-, named'
-, namedName
-, namedValue
+( Name
+-- * Constructors
+, gensym
+, name
+, nameI
+, formatName
+, isGenerated
+, __self
 ) where
 
-import Data.Function (on)
-import Data.String (IsString)
-import Data.Text (Text)
+import           Control.Effect.Fresh
+import           Data.Aeson
+import qualified Data.Char as Char
+import           Data.Hashable
+import           Data.String
+import           Data.Text (Text)
+import qualified Data.Text as Text
 
--- | User-specified and -relevant names.
-newtype Name = Name { unName :: Text }
-  deriving (Eq, IsString, Ord, Show)
+-- | The type of variable names.
+data Name
+  = Name Text
+  | I Int
+  deriving (Eq, Ord)
```

Comment:
I don’t think we should use this definition of `Name`, but I’m not sure that I have a good solution in mind for retaining the ability to `gensym` without it. My overall plan had been to _not_ use `gensym`, and instead pick reasonable names, secure in the knowledge that substitution is capture-avoiding because the name doesn’t appear in the terms, but even that isn’t quite true as evaluation of core (currently) substitutes the given name into the term when instantiating scopes like lambda bodies for evaluation.

@joshvera, IIRC @patrickt mentioned that you’re planning on relying on `gensym`; maybe we could chat about that purpose and see if there’s a better solution for either accommodating `gensym` or accomplishing the same goals without it?

Reply:
We could add another functor to the `Analysis` namespace:

```haskell
module Analysis.Functor.Gen where

data Gen a = I a | E a
```

and change the relevant analyses to operate over `Gen Name`? `Gen` is actually a lousy name because it’s taken by testing frameworks, but pretend I picked a good one for the sake of argument.